components/places/src/storage/history_metadata.rs

```diff
@@ -2,47 +2,209 @@
+                document_type: observation_with_title.document_type,
+                referrer_url: observation_with_title.referrer_url,
+            },
+            observation_with_title.title,
+        )
+    }
+}
+
+enum PlaceEntry {
+    Existing(i64),
+    CreateFor(Url, Option<String>),
+}
+
+impl PlaceEntry {
+    fn from(url: &str, tx: &PlacesTransaction<'_>, title: Option<String>) -> Result<Self> {
+        let url = Url::parse(url)?;
+        let place_id = tx.try_query_one(
+            "SELECT id FROM moz_places WHERE url_hash = hash(:url) AND url = :url",
+            &[(":url", &url.as_str())],
+            true,
+        )?;
+
+        Ok(match place_id {
+            Some(id) => PlaceEntry::Existing(id),
+            None => PlaceEntry::CreateFor(url, title),
+        })
+    }
+}
+
+trait DatabaseId {
+    fn get_or_insert(&self, tx: &PlacesTransaction<'_>) -> Result<i64>;
+}
+
+impl DatabaseId for PlaceEntry {
+    fn get_or_insert(&self, tx: &PlacesTransaction<'_>) -> Result<i64> {
+        Ok(match self {
+            PlaceEntry::Existing(id) => *id,
+            PlaceEntry::CreateFor(url, title) => {
+                let sql = "INSERT INTO moz_places (guid, url, title, url_hash)
+                VALUES (:guid, :url, :title, hash(:url))";
+
+                let guid = SyncGuid::random();
+
+                tx.execute_named_cached(
+                    sql,
+                    &[
+                        (":guid", &guid),
+                        (":title", &title),
+                        (":url", &url.as_str()),
+                    ],
+                )?;
+                tx.conn().last_insert_rowid()
+            }
+        })
+    }
+}
+
 enum SearchQueryEntry {
-    None,
     Existing(i64),
     CreateFor(String),
 }
 
-enum ParentDomainEntry {
-    None,
-    Existing(i64),
-    CreateFor(Url),
+impl DatabaseId for SearchQueryEntry {
+    fn get_or_insert(&self, tx: &PlacesTransaction<'_>) -> Result<i64> {
+        Ok(match self {
+            SearchQueryEntry::Existing(id) => *id,
+            SearchQueryEntry::CreateFor(term) => {
+                tx.execute_named_cached(
+                    "INSERT INTO moz_places_metadata_search_queries(term) VALUES (:term)",
+                    &[(":term", &term)],
+                )?;
+                tx.conn().last_insert_rowid()
+            }
+        })
+    }
+}
+
+impl SearchQueryEntry {
+    fn from(search_term: &str, tx: &PlacesTransaction<'_>) -> Result<Self> {
+        let lowercase_term = search_term.to_lowercase();
+        Ok(
+            match tx.try_query_one(
+                "SELECT id FROM moz_places_metadata_search_queries WHERE term = :term",
+                &[(":term", &lowercase_term)],
+                true,
+            )? {
+                Some(id) => SearchQueryEntry::Existing(id),
+                None => SearchQueryEntry::CreateFor(lowercase_term),
+            },
+        )
+    }
+}
+
+trait CanDebounce {
```

Comment:
nit: you don't need a trait for this, because no where do you use a generic type - ie, IIUC, you could just impl that function in each type and it would work fine. I'm fine if you think it adds some value though, but it also does add some cognitive overhead

Reply:
Cleaned this up, no more trait!